dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/alias/InclusionBasedPointerAnalysis.java

```diff
@@ -0,0 +1,897 @@
+
+    Initially, the graph has the following edges
+    - For every "store(A, V)" event, it has a "A -stores-> V" edge
+    - For every "R = load(A)" event, it has a "A -loads-> R" edge.
+    - For every assignment "R := B + k*x + o", we have a "B -(k*x+o)-> R" edge
+    For too complex assignments that do not match the structure, say "R := h(B1, B2, .., Bn)" where h is complex, we introduce edges "Bi -1x+0-> R".
+
+    On this initial graph, two rules are applied until saturation
+    (1) Transitivity: A -f_includes-> B AND B -g_includes-> C IMPLIES A -fg_includes-> C (notice that we need to compose the labels as well).
+    (2) Communication: if A -loads-> R and B -stores-> V and A and B can overlap (i.e. the corresponding Load/Store events can alias), then V -0x+0-> R (all values in V are included in R)
+
+    It can happen that these rules generate an edge A -f-> B although another edge A -g-> B already exists.
+    In this case, the edges get merged by a join operation.
+    To guarantee termination, the saturation of cyclic inclusion relationships get accelerated:
+    E.g. if A includes B (B -0x+0-> A) and B includes A+2 (A -0x+2-> B)
+    then A includes A+2x, B includes A+2x+2 and both include B+2x.
+
+    (*) We also generate special phi nodes that can improve precision on code that is not in SSA form.
+    (**) In the implementation we make an unsound assumption that x is non-negative.
+    This disallows negative (dynamic) indexing into arrays/pointers but gives additional precision in other cases.
+    Note that "o" can be negative which is sufficient to support "containerof" functionality.
+*/
+public class InclusionBasedPointerAnalysis implements AliasAnalysis {
+
+    private static final Logger logger = LogManager.getLogger(InclusionBasedPointerAnalysis.class);
+
+    private static final List<Integer> TOP = List.of(1);
+
+    // Debug: count inclusion tests grouped by complexity.
+    private final Map<Integer, int[]> totalAlignmentSizes = logger.isDebugEnabled() ? new HashMap<>() : null;
+    // Debug: count created variables.
+    private int totalVariables = 0;
+    // Debug: count variable substitutions.
+    private int totalReplacements = 0;
+    // Debug: count new edges already covered by existing ones.
+    private int failedAddInto = 0;
+    // Debug: count times a piece of new information was added to the graph.
+    private int succeededAddInto = 0;
+
+    // If enabled, the algorithm describes its internal graph to be written into a file.
+    private Graphviz graphviz;
+
+    // When a variable gains an includes-edge, it is added to this queue for later processing.
+    private final LinkedHashMap<Variable, List<Offset<Variable>>> queue = new LinkedHashMap<>();
+
+    // Maps memory events to variables representing their pointer set.
+    private final Map<MemoryCoreEvent, Offset<Variable>> addressVariables = new HashMap<>();
```

Comment:
The fact that the algorithm uses `Offset<Variable>` rather than `Variable` in many places is confusing.
Is this just a kind of optimization to avoid intermediate variables or something?